src/index/src/inverted_index/search/index_apply/predicates_apply.rs

```diff
@@ -0,0 +1,372 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use async_trait::async_trait;
+use common_base::BitVec;
+use greptime_proto::v1::index::InvertedIndexMetas;
+
+use super::SearchContext;
+use crate::inverted_index::error::{IndexNotFoundSnafu, Result};
+use crate::inverted_index::format::reader::InvertedIndexReader;
+use crate::inverted_index::search::fst_apply::{
+    FstApplier, IntersectionFstApplier, KeysFstApplier,
+};
+use crate::inverted_index::search::fst_values_mapper::FstValuesMapper;
+use crate::inverted_index::search::index_apply::IndexApplier;
+use crate::inverted_index::search::predicate::Predicate;
+
+type IndexName = String;
+
+/// `PredicatesIndexApplier` contains a collection of `FstApplier`s, each associated with an index name,
+/// to process and filter index data based on compiled predicates.
+pub struct PredicatesIndexApplier {
+    /// A list of `FstApplier`s, each associated with a specific index name
+    /// (e.g. a tag field uses its column name as index name)
+    fst_appliers: Vec<(IndexName, Box<dyn FstApplier>)>,
+
+    /// Options for reading an inverted index (e.g., how to handle missing index).
+    options: ReadOptions,
+}
+
+#[async_trait]
+impl IndexApplier for PredicatesIndexApplier {
+    /// Applies all `FstApplier`s to the data in the inverted index reader, intersecting the individual
+    /// bitmaps obtained for each index to result in a final set of indices.
+    async fn apply(
+        &self,
+        _context: SearchContext,
+        reader: &mut dyn InvertedIndexReader,
+    ) -> Result<Vec<usize>> {
+        let metadata = reader.metadata().await?;
+
+        let mut bitmap = Self::bitmap_full_range(&metadata);
+        // TODO(zhongzc): optimize the order of applying to make it quicker to return empty.
+        for (name, fst_applier) in &self.fst_appliers {
+            if bitmap.count_ones() == 0 {
+                break;
+            }
+
+            let Some(meta) = metadata.metas.get(name) else {
+                match self.options.index_not_found_strategy {
```

Comment:
Can we provide the options via the `SearchContext`?

Reply:
This is a very marginal decision. I think it depends on whether there will be an applier implementation that doesn't care about this configuration item in the future. If present, this change would be disruptive. But you know, I'm struggling. But now that you have this comment, I probably don't need to worry about it anymore, let's do it